core/includes/file.inc

```diff
@@ -2675,6 +2675,10 @@ function file_directory_temp() {
       // everything to use slash which is supported on all platforms.
       $temporary_directory = str_replace('\\', '/', $temporary_directory);
     }
+    
+    // Now when the temporary directory is found, we need to save it to
+    // the configuration file.
+    config_set('system.core', 'file_temporary_path', $temporary_directory);
```

Comment:
Setting the tmp dir shouldn't happen in function file_directory_temp() in file.inc, as that function is called several times, which means, the config item gets set again and again.

See: https://api.backdropcms.org/api/backdrop/core%21includes%21file.inc/function/file_directory_temp/1

It should rather go to core/modules/system/system.install to make sure, it only happens once on install.

Reply:
That's exactly why @jenlampton has put this comment https://github.com/backdrop/backdrop-issues/issues/4193#issuecomment-550064054 So we should call it only once - during the install - and all the other instances should replace with `config_get()`.

However, even if we didn't try to replace other calls to the function, I don't agree that it  

> gets set again and again

because there is a check in the very beginning of the module:

```
function file_directory_temp() {
  $temporary_directory = config_get('system.core', 'file_temporary_path');
  if (empty($temporary_directory)) {
```

In other words, the execution never proceeds if temp directory has already been put into the config file. So nothing terrible in case of multiple calls too.

Reply:
> So nothing terrible in case of multiple calls too.

True, but there's no need to do it here. Running config_set once at install would be much cleaner. My recommendation is still to _not_ do this in file.inc but in the installer.

Reply:
Well, I'll be more happy if core gets more optimized. The multiple checks have already been there. They are not cause by my PR. 

My PR is trying to deal with specific problem that Backdrop UI is giving wrong information to user. But let's see how this goes - if core maintainers would be willing to introduce multiple changes instead one suggested.

Reply:
> the whole function needs to move to installer.

Function file_directory_temp() is a utility and at its perfect place.
Please read at least the first few sentences of the api description of function file_directory_temp().

Doing the _initial config_set_ of the tmp dir should go to the installer.

Reply:
> > the whole function needs to move to installer.
> 
> Function file_directory_temp() is a utility and at its perfect place.
> Please read at least the first few sentences of the api description of function file_directory_temp().
> 
> Doing the _initial config_set_ of the tmp dir should go to the installer.

@indigoxela The above discussion is little outdated, because after you explained your idea on gitter I had created second PR just like you wanted.

Reply:
The reason I thought this PR was okay was because, it makes a pretty small change and I checked D7 version and it does the same thing: saves the constructed variable if it's not already saved. It should only happen once no matter how many times the function is called. (Putting it in system.install may be cleaner but perhaps this PR helps in case where someone manually deletes the variable so it can automatically recreate it.)

And that's the only change I think is needed: I don't think we should try to replace the use of this function. (This old d.org issue https://www.drupal.org/node/930122 describes that they intentionally used this function instead of calling the variable directly.)

Reply:
> Putting it in system.install may be cleaner but perhaps this PR helps in case where someone manually deletes the variable so it can automatically recreate it.

Interesting catch. There will be some folks who get dirty their hands deleting entries in the config file. Then we are back to the situation when UI shows a path that is not present in the config file. So my first PR might still be better than the second one.